README.md
# data-visualizer

[![GitHub Downloads][github-img]][github-url]
[![Travis-CI][travis-img]][travis-url]
[![Coveralls][coveralls-img]][coveralls-url]
[![Code Climate][codeclimate-img]][codeclimate-url]
[![Code Climate][codeclimate-issues-img]][codeclimate-issues-url]
[![js-standard-style][standard-img]][standard-url]

Visualize data

## Installation

``` bash
  $ npm install 'bagrounds/data-visualizer'
```

## [Documentation][gh-pages-url]


## Usage
...

## Tests
``` bash
  $ npm test
```


## [Changelog][changelog-url]

## License
[MIT][license-url]


[changelog-url]: CHANGELOG.md

[license-url]: LICENSE

[standard-img]: https://img.shields.io/badge/code%20style-standard-brightgreen.svg
[standard-url]: http://standardjs.com/

[github-img]: https://img.shields.io/github/downloads/bagrounds/data-visualizer/total.svg
[github-url]: https://github.com/bagrounds/data-visualizer

[travis-img]: https://img.shields.io/travis/bagrounds/data-visualizer/master.svg
[travis-url]: https://travis-ci.org/bagrounds/data-visualizer

[coveralls-img]: https://coveralls.io/repos/github/bagrounds/data-visualizer/badge.svg?branch=master
[coveralls-url]: https://coveralls.io/github/bagrounds/data-visualizer?branch=master

[codeclimate-img]: https://codeclimate.com/github/bagrounds/data-visualizer/badges/gpa.svg
[codeclimate-url]: https://codeclimate.com/github/bagrounds/data-visualizer

[codeclimate-issues-img]: https://codeclimate.com/github/bagrounds/data-visualizer/badges/issue_count.svg
[codeclimate-issues-url]: https://codeclimate.com/github/bagrounds/data-visualizer/issues

[gh-pages-url]: http://bagrounds.github.io/data-visualizer


CHANGELOG.md
## Change Log

All notable changes to this project will be documented in this section. This
project adheres to [Semantic Versioning](http://semver.org/).

Please use: Added, Changed, Deprecated, Removed, Fixed, or Security as labels.
<br>

#### 0.3.0 - (Unreleased)
##### Added
* ...

##### Changed
* ...

##### Deprecated
* ...

##### Removed
* ...

##### Fixed
* ...

##### Security
* ...

#### 0.2.0 - (2016-08-16)
##### Added
* support for line charts

#### 0.1.0 - (2016-06-21)
##### Added
* All files from bagrounds/template-spa v0.2.0
* bar-chart module from my external project

##### Changed
* relevant files from template to suit this project
* CHANGELOG (this) to suit this project
* dependencies as appropriate

##### Removed
* irrelevant files



test/test-data-visualizer.js
/**
 * Tests for data-visualizer
 */
;(function () {
  /* global describe, it */
  'use strict'

  /* Imports */
  var expect = require('chai').expect

  var dataVisualizer = require('../data-visualizer')

  /* Tests */
  describe('data-visualizer', function () {
    describe('stub', function () {
      it('should pass', function () {
        expect(true).to.be.true
        expect(dataVisualizer).to.be.a.function
      })
    })
  })
})()


js/bar-chart.js
/**
 *
 * @module chart
 */
;(function () {
  'use strict'
  /* global document */

  var React = require('react')
  var ReactDOM = require('react-dom')

  var BarChart = require('react-d3-basic').BarChart

  module.exports = chart

  /**
   * Plots a bar chart
   *
   * @function plotBars
   *
   * @alias chart
   *
   * @param {Object} options all function parameters
   * @param {Array<Object>} options.data properly formatted dataSet
   * @param {Number} options.width chart width
   * @param {Number} options.height chart height
   * @param {String} options.xValueId field name for x axis value
   * @param {String} options.xLabel x axis label
   * @param {String} options.yValueId field name for y axis value
   * @param {String} options.yLabel y axis label
   * @param {String} options.title chart title
   * @param {String} options.parentDiv where to render the chart
   *
   * @param {Function} callback handle results
   */
  function chart (options, callback) {
    var error = invalidOptions(options)

    if (error) {
      callback(error)
      return
    }

    var xScale = 'ordinal'
    var yTicks = [10]

    function width () {
      if (options.width) {
        return options.width
      }

      var parent = document.getElementById(options.parentDiv)

      var width = parent.offsetWidth

      return width
    }

    function height () {
      return options.height
        ? options.height
        : undefined
    }
    // var height = 400

    ReactDOM.render(React.createElement(BarChart, {
      title: options.title,
      data: options.data,
      width: width(),
      height: height(),
      margins: {
        top: 20,
        bottom: 200,
        left: 80,
        right: 10
      },
      chartSeries: [
        {
          field: options.yValueId,
          name: options.yLabel
        }
      ],
      x: function (d) {
        return d[options.xValueId]
      },
      xScale: xScale,
      yTicks: yTicks
    }), document.getElementById(options.parentDiv))

    callback()
  }

  /**
   *
   * TODO: actually check stuff
   *
   * @param {Object} options contains all function parameters
   * @param {Array<Object>} options.data properly formatted dataSet
   * @param {Number} options.width chart width
   * @param {Number} options.height chart height
   * @param {String} options.xValueId field name for x axis value
   * @param {String} options.xLabel x axis label
   * @param {String} options.yValueId field name for y axis value
   * @param {String} options.yLabel y axis label
   * @param {String} options.title chart title
   * @param {String} options.parentDiv where to render the chart
   *
   * @returns {Error|Boolean} an error if problems, or false
   */
  function invalidOptions (options) {
    var requiredOptions = [
      'data',
      'width',
      'height',
      'xValueId',
      'yValueId',
      'parentDiv'
    ]

    var error

    requiredOptions.forEach(function (option) {
      if (!options[option]) {
        var message = 'option: ' + option + ' missing!'
        error = new Error(message)
      }
    })

    return error
  }
})()


data-visualizer.js
/**
 * Visualize data
 *
 * @module data-visualizer
 */
;(function () {
  'use strict'

  /* imports */
  var typeCheck = require('type-check').typeCheck
  var barChart = require('./js/chart')

  /* exports */
  module.exports = visualize

  /**
   *
   * @function visualize
   *
   * @param {Object} options all function parameters
   * @param {Function} callback handle results
   */
  function visualize (options, callback) {
    var error = invalidOptions(options)

    if (error) {
      callback(error)
      console.error(error)
      return
    }

    barChart(options, callback)
  }

  /**
   * Validate options
   *
   * @param {Object} options that were passed to visualize
   * @return {Error} if options are invalid
   */
  function invalidOptions (options) {
    if (!typeCheck('Object', options)) {
      var message = 'options should be an object'
      return new Error(message)
    }

    return null
  }
})()


js/chart.js
/**
 *
 * @module chart
 */
;(function () {
  'use strict'
  /* global document */

  var React = require('react')
  var ReactDOM = require('react-dom')

  var BarChart = require('react-d3-basic').BarChart
  var LineChart= require('react-d3-basic').LineChart

  var CHART_TYPES = {
    line: LineChart,
    bar: BarChart
  }

  module.exports = chart

  /**
   * Plots a bar chart
   *
   * @function plotBars
   *
   * @alias chart
   *
   * @param {Object} options all function parameters
   * @param {Array<Object>} options.data properly formatted dataSet
   * @param {Number} options.width chart width
   * @param {Number} options.height chart height
   * @param {String} options.xValueId field name for x axis value
   * @param {String} options.xLabel x axis label
   * @param {String} options.yValueId field name for y axis value
   * @param {String} options.yLabel y axis label
   * @param {String} options.title chart title
   * @param {String} options.parentDiv where to render the chart
   * @param {String} options.type line or bar (default)
   *
   * @param {Function} callback handle results
   */
  function chart (options, callback) {
    var error = invalidOptions(options)

    if (error) {
      callback(error)
      return
    }

    if (!options.type) {
      options.type = 'bar'
    }

    var xScale = 'ordinal'
    var yTicks = [10]

    function width () {
      if (options.width) {
        return options.width
      }

      var parent = document.getElementById(options.parentDiv)

      var width = parent.offsetWidth

      return width
    }

    function height () {
      return options.height
        ? options.height
        : undefined
    }
    // var height = 400

    ReactDOM.render(React.createElement(CHART_TYPES[options.type], {
      title: options.title,
      data: options.data,
      width: width(),
      height: height(),
      margins: {
        top: 20,
        bottom: 200,
        left: 80,
        right: 10
      },
      chartSeries: [
        {
          field: options.yValueId,
          name: options.yLabel
        }
      ],
      x: function (d) {
        return d[options.xValueId]
      },
      xScale: xScale,
      yTicks: yTicks
    }), document.getElementById(options.parentDiv))

    callback()
  }

  /**
   *
   * TODO: actually check stuff
   *
   * @param {Object} options contains all function parameters
   * @param {Array<Object>} options.data properly formatted dataSet
   * @param {Number} options.width chart width
   * @param {Number} options.height chart height
   * @param {String} options.xValueId field name for x axis value
   * @param {String} options.xLabel x axis label
   * @param {String} options.yValueId field name for y axis value
   * @param {String} options.yLabel y axis label
   * @param {String} options.title chart title
   * @param {String} options.parentDiv where to render the chart
   *
   * @returns {Error|Boolean} an error if problems, or false
   */
  function invalidOptions (options) {
    var requiredOptions = [
      'data',
      'width',
      'height',
      'xValueId',
      'yValueId',
      'parentDiv'
    ]

    var error

    requiredOptions.forEach(function (option) {
      if (!options[option]) {
        var message = 'option: ' + option + ' missing!'
        error = new Error(message)
      }
    })

    return error
  }
})()
